rewrite-java/src/main/java/org/openrewrite/java/cleanup/FinalizePrivateFields.java

```diff
@@ -0,0 +1,240 @@
+                        v -> v.getInitializer() != null ? 1 : 0));
+
+                CollectPrivateFieldsAssignmentCounts.collect(classDecl, privateFieldAssignCountMap);
+
+                privateFieldsToBeFinalized = privateFieldAssignCountMap.entrySet()
+                    .stream()
+                    .filter(entry -> entry.getValue() == 1)
+                    .map(Map.Entry::getKey)
+                    .collect(Collectors.toSet());
+
+                return super.visitClassDeclaration(classDecl, ctx);
+            }
+
+            @Override
+            public J.VariableDeclarations visitVariableDeclarations(J.VariableDeclarations multiVariable,
+                ExecutionContext ctx) {
+                J.VariableDeclarations mv = super.visitVariableDeclarations(multiVariable, ctx);
+
+                boolean canAllVariablesBeFinalized = mv.getVariables()
+                    .stream()
+                    .map(J.VariableDeclarations.NamedVariable::getVariableType)
+                    .allMatch(v -> privateFieldsToBeFinalized.contains(v));
+
+                if (canAllVariablesBeFinalized) {
+                    mv = autoFormat(mv.withModifiers(ListUtils.concat(mv.getModifiers(),
+                        new J.Modifier(Tree.randomId(), Space.EMPTY, Markers.EMPTY, J.Modifier.Type.Final,
+                            Collections.emptyList()))), ctx);
+                }
+
+                return mv;
+            }
+        };
+    }
+
+    /**
+     * Collect private and non-final fields from a class
+     */
+    private static List<J.VariableDeclarations.NamedVariable> collectPrivateFields(J.ClassDeclaration classDecl) {
+        return classDecl.getBody()
+            .getStatements()
+            .stream()
+            .filter(statement -> statement instanceof J.VariableDeclarations)
+            .map(J.VariableDeclarations.class::cast)
+            .filter(mv -> mv.hasModifier(J.Modifier.Type.Private) && !mv.hasModifier(J.Modifier.Type.Final))
+            .map(J.VariableDeclarations::getVariables)
+            .flatMap(Collection::stream)
+            .collect(Collectors.toList());
+    }
+
+    private static class CollectPrivateFieldsAssignmentCounts extends JavaIsoVisitor<Map<JavaType.Variable, Integer>> {
+
+        /**
+         * Collects private fields assignment counts, count rules are:
+         * (1) any assignment in class constructor, instance variable initializer or initializer block count for 1.
+         * (2) any assignment in other place count for 2.
+         * (3) any assignment in a loop in anywhere count for 2.
+         * By giving 3 rules above, if a private field has an assigned count equal to 1, it should be finalized.
+         *
+         * @param j The subtree to search, supposed to be at Class declaration level.
+         */
+        static void collect(J j, Map<JavaType.Variable, Integer> assignedCountMap) {
+            new CollectPrivateFieldsAssignmentCounts().visit(j, assignedCountMap);
+        }
```

Comment:
`TreeVisitor.collect()` provides a shortcut/generalization of this pattern. I haven't checked that this compiles, but something like this:
```suggestion
        static Map<JavaType.Variable, Integer> collect(J j) {
             return TreeVisitor.collect(new CollectPrivateFieldsAssignmentCounts(), j, new HashMap<>());
        }
```